pkg/hub/consul_health.go

```diff
@@ -0,0 +1,181 @@
+package hub
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+)
+
+var (
+	// The maximum time to run a blocking query to the catalog for.
+	DefaultRefreshTime = 30 * time.Second
+
+	// How long to configured the consul TTL check with. This is the maximum time
+	// the service has to update it's TTL again before it's marked as failed.
+	CheckTTL = "30s"
+
+	// ConsulService is the service name that is used to identify hubs. Each
+	// instance of ConsulHealth identifies itself as a separate instance of this service.
+	ConsulService = "hub"
+)
+
+// ConsulHealth advertises an instance of the hub service to consul and also
+// monitors the service catalog for other instances. It provides the ability for
+// each hub to know which other hubs are currently available to send traffic to.
+type ConsulHealth struct {
+	api *consul.Client
+	id  string
+
+	mu   sync.RWMutex
+	hubs map[string]struct{}
+}
+
+// NewConsulHealth creates a ConsulHealth instance, identified by id.
+func NewConsulHealth(id string, cfg *consul.Config) (*ConsulHealth, error) {
+	c, err := consul.NewClient(cfg)
```

Comment:
Do you want to make a new client here or do you want to pass in the existing client you already made earlier?

Reply:
Yeah, I do. The clients are so lightweight because they share a connection pool and this is the better API imho.

Reply:
For testing, the docker-compose file already has consul setup too so along with all the other dependencies anyway.

Reply:
Sounds good, the reason I mentioned the helper is because it'll make sure you have a clean Consul with no data, and tear it down. You may need to do that sort of stuff for the docker-compose one.

Reply:
Yeah, the docker-compose one runs in dev mode and the tests cleanup after themselves, so at worst you just have to bounce the docker-compose services.